### PR TITLE
Refactor request

### PR DESCRIPTION
**Pull request by anshumansworld**
This patch contains following changes:
* Removed 3rd party library (Volley) for network requests
* Added new package network that will be responsible for network calls
* Some minor optimization in PubnativeRequest class

### CODE REVIEW

File: library/src/main/java/net/pubnative/library/request/PubnativeRequest.java [hunk 1]
```diff
@@ -285,9 +282,7 @@ protected void sendNetworkRequest() {
 
         } else {
 
-            RequestQueue queue = Volley.newRequestQueue(this.context);
-            StringRequest request = new StringRequest(Request.Method.GET, url, this, this);
-            queue.add(request);
+            RequestManager.sendRequest(new Request(Request.Method.GET, url, this, this));
```
Comment (cerberillo) [inline library/src/main/java/net/pubnative/library/request/PubnativeRequest.java]: It will look better like this:

`Request.send(METHOD, URL, LISTENER)`

So, simplify listener with 2 callbacks and make the `Request` object send itself to the `RequestManager`

**Comment by cerberillo** [comment]
Test changes

* Please review tests also, there are Volley naming and some of them don't make sense.
* New Request queue should also be tested somehow.
* We need to add callback testing to the request, so we know that when called with null the invoke methods are not crashing

File: library/src/main/java/net/pubnative/library/network/Request.java [hunk 2]
```diff
@@ -0,0 +1,76 @@
...
+    }
+
+    public Response.ErrorListener getErrorListener() {
+
+        return mErrorListener;
+    }
+
+    public void deliverResponse(String response) {
+
+        mListener.onResponse(response);
```
Comment (cerberillo) [inline library/src/main/java/net/pubnative/library/network/Request.java]: You should check listener nullity

File: library/src/main/java/net/pubnative/library/network/RequestTask.java [hunk 3]
```diff
@@ -0,0 +1,99 @@
...
+            mExecuteRequestRunnable = new Runnable() {
+
+                @Override
+                public void run() {
+                    executeRequest();
+                }
+            };
+        }
+        return mExecuteRequestRunnable;
+    }
```
Comment (cerberillo) [inline library/src/main/java/net/pubnative/library/network/RequestTask.java]: If this is only being used at one point, wouldn't it make sense to create the runnable there at runtime?

File: library/src/main/java/net/pubnative/library/network/Response.java [hunk 4]
```diff
@@ -0,0 +1,57 @@
...
+import java.io.IOException;
+import java.io.InputStream;
+
+/**
+ * Created by jaiswal.anshuman on 2/2/2016.
+ */
+public class Response {
+
+    public String result;
+    public Exception error;
```
Comment (cerberillo) [inline library/src/main/java/net/pubnative/library/network/Response.java]: We should use only 1 naming convention 
Shouldn't we use mResult and mError also?

File: library/src/main/java/net/pubnative/library/network/Response.java [hunk 5]
```diff
@@ -0,0 +1,57 @@
...
+
+    public void setResult(String result) {
+        this.result = result;
+    }
+
+    public void setResult(InputStream is) {
+
+        BufferedInputStream bis = new BufferedInputStream(is);
+        ByteArrayOutputStream buf = new ByteArrayOutputStream();
```
Comment (cerberillo) [inline library/src/main/java/net/pubnative/library/network/Response.java]: Please let's use long variable naming `outputStream` and `inputStream` better than `bis` and `buf` that break readability

File: library/src/main/java/net/pubnative/library/network/Response.java [hunk 6]
```diff
@@ -0,0 +1,57 @@
...
+        this.error = error;
+    }
+
+    public interface ErrorListener {
+        void onErrorResponse(Exception error);
+    }
+
+    public interface Listener {
+        void onResponse(String response);
+    }
```
Comment (cerberillo) [inline library/src/main/java/net/pubnative/library/network/Response.java]: Let's unity the listeners into one and let's return the request with it, so we can compare objects in case you launch several of them at the same time

File: library/src/main/java/net/pubnative/library/network/RequestTask.java [hunk 7]
```diff
@@ -0,0 +1,99 @@
...
+
+            connection.connect();
+
+            int responseCode = connection.getResponseCode();
+
+            if(responseCode == -1) {
+                throw new IOException("Could not retrieve response code from HttpUrlConnection.");
+            } else {
+                response.setResult(connection.getInputStream());
+            }
```
Comment (cerberillo) [inline library/src/main/java/net/pubnative/library/network/RequestTask.java]: The only admitted response code should be 200 any other ay be considered as an error

Comment (cerberillo) [inline library/src/main/java/net/pubnative/library/network/RequestTask.java]: It's in fact `HTTP_OK`

File: library/src/main/java/net/pubnative/library/network/RequestTask.java (hunk 3 above)
Comment (anshumansworld) [inline library/src/main/java/net/pubnative/library/network/RequestTask.java]: If we create runnable, it couldn't be reused.
The idea is:
A requestTask is being retrieved from queue of RequestTask by using method poll() and the runnable from task so runnable is attached to task

File: library/src/main/java/net/pubnative/library/network/Response.java (hunk 4 above)
Comment (anshumansworld) [inline library/src/main/java/net/pubnative/library/network/Response.java]: (y)

File: library/src/main/java/net/pubnative/library/request/PubnativeRequest.java (hunk 1 above)
Comment (anshumansworld) [inline library/src/main/java/net/pubnative/library/request/PubnativeRequest.java]: :+1:

File: library/src/main/java/net/pubnative/library/network/Request.java (hunk 2 above)
Comment (anshumansworld) [inline library/src/main/java/net/pubnative/library/network/Request.java]: :+1:

File: library/src/main/java/net/pubnative/library/network/Response.java (hunk 5 above)
Comment (anshumansworld) [inline library/src/main/java/net/pubnative/library/network/Response.java]: :+1:

File: library/src/main/java/net/pubnative/library/network/Response.java (hunk 6 above)
Comment (anshumansworld) [inline library/src/main/java/net/pubnative/library/network/Response.java]: :+1:

File: library/src/main/java/net/pubnative/library/network/RequestTask.java (hunk 7 above)
Comment (anshumansworld) [inline library/src/main/java/net/pubnative/library/network/RequestTask.java]: :+1:

File: library/src/main/java/net/pubnative/library/network/PubnativeAPIRequest.java [hunk 8]
```diff
@@ -0,0 +1,69 @@
...
+
+import java.net.MalformedURLException;
+import java.net.URL;
+
+public class PubnativeAPIRequest {
+
+    private static String TAG = PubnativeAPIRequest.class.getSimpleName();
+
+    //timeout in ms
+    public static int TIME_OUT = 10000;
```
Comment (cerberillo) [inline library/src/main/java/net/pubnative/library/network/PubnativeAPIRequest.java]: We should keep timeout of 3 seconds instead of 10 which is too much

File: library/src/main/java/net/pubnative/library/network/PubnativeAPIRequest.java [hunk 9]
```diff
@@ -0,0 +1,69 @@
...
+    }
+
+    public PubnativeAPIResponse.Listener getListener() {
+        return mListener;
+    }
+
+    public void deliverResponse(String response) {
+
+        if(mListener != null)
+            mListener.onResponse(response);
```
Comment (cerberillo) [inline library/src/main/java/net/pubnative/library/network/PubnativeAPIRequest.java]: Let's add curly brackets even if it's one line only

File: library/src/main/java/net/pubnative/library/network/PubnativeAPIRequestTask.java [hunk 10]
```diff
@@ -0,0 +1,103 @@
...
+            connection.setConnectTimeout(PubnativeAPIRequest.TIME_OUT);
+            connection.setReadTimeout(PubnativeAPIRequest.TIME_OUT);
+            connection.setRequestMethod("GET");
+            connection.setDoInput(true);
+
+            connection.connect();
+
+            int responseCode = connection.getResponseCode();
+
+            if(responseCode == HttpStatus.SC_OK) {
```
Comment (cerberillo) [inline library/src/main/java/net/pubnative/library/network/PubnativeAPIRequestTask.java]: Instead of using apache HttpStatus we should use HTTP_OK which is OS dependant

File: library/src/main/java/net/pubnative/library/network/PubnativeAPIRequestTask.java [hunk 11]
```diff
@@ -0,0 +1,103 @@
...
+            } else {
+                throw new IOException("Could not retrieve response code from HttpUrlConnection.");
+            }
+        } catch (IOException e) {
+
+            Log.e(TAG, e.getMessage());
+            pubnativeAPIResponse.setError(e);
+        } finally {
+
+            this.mResponsePoster.execute(new ResponseDeliveryRunnable(mPubnativeAPIRequest, pubnativeAPIResponse));
```
Comment (cerberillo) [inline library/src/main/java/net/pubnative/library/network/PubnativeAPIRequestTask.java]: Use of this with `m-named` variables looks redundant, pelase remove this from those that are  `m-named`

Comment (cerberillo) [inline library/src/main/java/net/pubnative/library/network/PubnativeAPIRequestTask.java]: Wouldn't be better to declare the runnable directly inside?

```java
mResponsePoster.execute(new Runnable() {
  @Override
  public void run() {
    if(this.mPubnativeAPIResponse.isSuccess()) {
      this.mPubnativeAPIRequest.deliverResponse(this.mPubnativeAPIResponse.mResult);
    } else {
      this.mPubnativeAPIRequest.deliverError(this.mPubnativeAPIResponse.mError);
    }
  }
});
```

File: library/src/main/java/net/pubnative/library/network/PubnativeAPIResponse.java [hunk 12]
```diff
@@ -0,0 +1,57 @@
...
+public class PubnativeAPIResponse {
+
+    private static String TAG = PubnativeAPIResponse.class.getSimpleName();
+
+    public String mResult;
+    public Exception mError;
+
+    public void setResult(String result) {
+        mResult = result;
+    }
```
Comment (cerberillo) [inline library/src/main/java/net/pubnative/library/network/PubnativeAPIResponse.java]: Creatting setters with public attributes seems redundant, wether use protected attributes and public setters/getters or public attributes without setters/getters

File: library/src/main/java/net/pubnative/library/network/PubnativeAPIResponse.java [hunk 13]
```diff
@@ -0,0 +1,57 @@
...
+
+        try {
+
+            int result = inputStream.read();
+            while(result != -1) {
+                byte byteResult = (byte)result;
+                outputStream.write(byteResult);
+                result = inputStream.read();
+            }
+            mResult = outputStream.toString();
```
Comment (cerberillo) [inline library/src/main/java/net/pubnative/library/network/PubnativeAPIResponse.java]: You should reuse setResult in this case

File: library/src/main/java/net/pubnative/library/request/PubnativeRequest.java [hunk 14]
```diff
@@ -31,22 +27,22 @@
...
-    private static final   String NATIVE_TYPE_URL = "native";
+    protected static final String   BASE_URL        = "http://api.pubnative.net/api/partner/v2/promotions";
+    private static final   String   NATIVE_TYPE_URL = "native";
 
-    protected Context             context;
-    protected Endpoint            endpoint;
-    protected Map<String, String> requestParameters = new HashMap<String, String>();;
-    protected Listener            listener;
+    protected Context               mContext;
+    protected Type                  type;
```
Comment (cerberillo) [inline library/src/main/java/net/pubnative/library/request/PubnativeRequest.java]: why type is not m-named

File: library/src/main/java/net/pubnative/library/request/PubnativeRequest.java [hunk 15]
```diff
@@ -103,7 +98,14 @@
          * @param request Request object used for making the request
          * @param ex      Exception that caused the failure
          */
-        void onPubnativeRequestFail(PubnativeRequest request, Exception ex);
+        void onRequestFailed(PubnativeRequest request, Exception ex);
```
Comment (cerberillo) [inline library/src/main/java/net/pubnative/library/request/PubnativeRequest.java]: Please, keep the naming for PubnativeRequest.Listener and the method names for public use classes

**Comment by cerberillo** [comment]
We should add some callbacks over the `invoke` methods of the request to ensure that they don't crash when calling them with null listener and that the listener callback is called when is different than null

File: library/src/main/java/net/pubnative/library/network/PubnativeAPIRequest.java (hunk 8 above)
Comment (anshumansworld) [inline library/src/main/java/net/pubnative/library/network/PubnativeAPIRequest.java]: :+1:

File: library/src/main/java/net/pubnative/library/network/PubnativeAPIRequest.java (hunk 9 above)
Comment (anshumansworld) [inline library/src/main/java/net/pubnative/library/network/PubnativeAPIRequest.java]: :+1:

File: library/src/main/java/net/pubnative/library/network/PubnativeAPIResponse.java (hunk 12 above)
Comment (anshumansworld) [inline library/src/main/java/net/pubnative/library/network/PubnativeAPIResponse.java]: :+1:

File: library/src/main/java/net/pubnative/library/network/PubnativeAPIRequestTask.java (hunk 10 above)
Comment (anshumansworld) [inline library/src/main/java/net/pubnative/library/network/PubnativeAPIRequestTask.java]: :+1:

File: library/src/main/java/net/pubnative/library/network/PubnativeAPIRequestTask.java (hunk 11 above)
Comment (anshumansworld) [inline library/src/main/java/net/pubnative/library/network/PubnativeAPIRequestTask.java]: :+1:

File: library/src/main/java/net/pubnative/library/request/PubnativeRequest.java (hunk 15 above)
Comment (anshumansworld) [inline library/src/main/java/net/pubnative/library/request/PubnativeRequest.java]: :+1:

File: library/src/main/java/net/pubnative/library/network/PubnativeAPIResponse.java (hunk 13 above)
Comment (anshumansworld) [inline library/src/main/java/net/pubnative/library/network/PubnativeAPIResponse.java]: :+1:

File: library/src/main/java/net/pubnative/library/request/PubnativeRequest.java (hunk 14 above)
Comment (anshumansworld) [inline library/src/main/java/net/pubnative/library/request/PubnativeRequest.java]: :+1:

File: library/src/main/java/net/pubnative/library/network/PubnativeAPIRequest.java [hunk 16]
```diff
@@ -0,0 +1,71 @@
...
+        }
+    }
+
+    public void deliverError(Exception error) {
+
+        if(mListener != null) {
+            mListener.onErrorResponse(error);
+        }
+    }
```
Comment (cerberillo) [inline library/src/main/java/net/pubnative/library/network/PubnativeAPIRequest.java]: in order to be cross-platform consistent, we're using "invoke<CallbackMethod>" with callback listeners

File: library/src/main/java/net/pubnative/library/network/PubnativeAPIRequestManager.java [hunk 17]
```diff
@@ -0,0 +1,129 @@
...
+    }
+
+    /**
+     * Starts a pubnativeAPIRequest
+     */
+    public static void sendRequest(PubnativeAPIRequest pubnativeAPIRequest) {
+    	/*
+         * Gets a task from the pool of tasks, returning null if the pool is empty
+         */
+        PubnativeAPIRequestTask pubnativeAPIRequestTask = sInstance.mPubnativeAPIRequestTaskWorkQueue.poll();
```
Comment (cerberillo) [inline library/src/main/java/net/pubnative/library/network/PubnativeAPIRequestManager.java]: we should use the static accessor if described

File: library/src/main/java/net/pubnative/library/network/PubnativeAPIResponse.java [hunk 18]
```diff
@@ -0,0 +1,65 @@
...
+
+    public void setError(Exception error) {
+        mError = error;
+    }
+
+    public interface Listener {
+        void onResponse(String response);
+
+        void onErrorResponse(Exception error);
+    }
```
Comment (cerberillo) [inline library/src/main/java/net/pubnative/library/network/PubnativeAPIResponse.java]: Why should the Listener be declared in the Response if it's being accessed through the Request and the safe acessors are there also?

File: library/src/main/java/net/pubnative/library/request/PubnativeRequest.java [hunk 19]
```diff
@@ -31,22 +27,22 @@
...
+    protected Listener              mListener;
 
     /**
      * These are the various types of adds pubnative support
      */
-    public enum Endpoint {
+    public enum Type {
 
         NATIVE,
     }
```
Comment (cerberillo) [inline library/src/main/java/net/pubnative/library/request/PubnativeRequest.java]: Please, leave this as Endpoint

File: library/src/test/java/net/pubnative/library/request/PubnativeRequestTest.java [hunk 20]
```diff
@@ -206,24 +206,63 @@ public void testOnResponseWithNullData() {
...
+        Exception               error    = mock(Exception.class);
+        request.invokeOnPubnativeRequestFailure(error);
+
+        verify(listener, times(0)).onPubnativeRequestFailed(eq(request), eq(error));
+
+        APIRequestResponseModel reponseModel = mock(APIRequestResponseModel.class);
+        reponseModel.ads = new ArrayList<PubnativeAdModel>();
+        request.invokeOnPubnativeRequestSuccess(reponseModel.ads);
+
+        verify(listener, times(0)).onPubnativeRequestFailed(eq(request), eq(error));
```
Comment (cerberillo) [inline library/src/test/java/net/pubnative/library/request/PubnativeRequestTest.java]: `PubnativeRequest.Listener` invokes should be verified in a different file on the `PubnativeRequest` scope named `PubnativeRequest.ListenerTest`

And since this ins unit testing, one method per-test please we should avoid multiple tests inside one test

File: library/src/main/java/net/pubnative/library/network/PubnativeAPIRequest.java (hunk 16 above)
Comment (anshumansworld) [inline library/src/main/java/net/pubnative/library/network/PubnativeAPIRequest.java]: :+1:

File: library/src/main/java/net/pubnative/library/network/PubnativeAPIRequestManager.java (hunk 17 above)
Comment (anshumansworld) [inline library/src/main/java/net/pubnative/library/network/PubnativeAPIRequestManager.java]: :+1:

File: library/src/main/java/net/pubnative/library/request/PubnativeRequest.java (hunk 19 above)
Comment (anshumansworld) [inline library/src/main/java/net/pubnative/library/request/PubnativeRequest.java]: :+1:

File: library/src/main/java/net/pubnative/library/network/PubnativeAPIResponse.java (hunk 18 above)
Comment (anshumansworld) [inline library/src/main/java/net/pubnative/library/network/PubnativeAPIResponse.java]: :+1:

File: library/src/test/java/net/pubnative/library/request/PubnativeRequestTest.java (hunk 20 above)
Comment (anshumansworld) [inline library/src/test/java/net/pubnative/library/request/PubnativeRequestTest.java]: :+1:

File: library/src/main/java/net/pubnative/library/network/PubnativeAPIRequest.java [hunk 21]
```diff
@@ -0,0 +1,77 @@
...
+        mMethod = method;
+        mUrl = url;
+        mListener = listener;
+    }
+
+    public enum Method {
+        GET
+    }
+
+    public interface APIRequestListener {
```
Comment (cerberillo) [inline library/src/main/java/net/pubnative/library/network/PubnativeAPIRequest.java]: rename `PubnativeAPIRequest.APIRequestListener` to `PubnativeAPIRequest.Listener` would make more sense

File: library/src/main/java/net/pubnative/library/network/PubnativeAPIRequestTask.java [hunk 22]
```diff
@@ -0,0 +1,91 @@
...
+            connection.setDoInput(true);
+
+            connection.connect();
+
+            int responseCode = connection.getResponseCode();
+
+            if(responseCode == HttpURLConnection.HTTP_OK) {
+                pubnativeAPIResponse.setResult(connection.getInputStream());
+            } else {
+                throw new IOException("Could not retrieve response code from HttpUrlConnection.");
```
Comment (cerberillo) [inline library/src/main/java/net/pubnative/library/network/PubnativeAPIRequestTask.java]: This error should be something like `"Server error: " + responseCode` so we can see it in the returned error

File: library/src/main/java/net/pubnative/library/network/PubnativeAPIRequestTask.java [hunk 23]
```diff
@@ -0,0 +1,91 @@
...
+            mResponsePoster.execute(new Runnable() {
+
+                @Override
+                public void run() {
+                    if(pubnativeAPIResponse.isSuccess()) {
+                        mPubnativeAPIRequest.deliverResponse(pubnativeAPIResponse.getResult());
+                    } else {
+                        mPubnativeAPIRequest.deliverError(pubnativeAPIResponse.getError());
+                    }
+                    PubnativeAPIRequestManager.getInstance().recycleTask(PubnativeAPIRequestTask.this);
```
Comment (cerberillo) [inline library/src/main/java/net/pubnative/library/network/PubnativeAPIRequestTask.java]: What if we aprocess all this inside a method `processResponse`?? It could make all this more readable

File: library/src/main/java/net/pubnative/library/request/PubnativeRequest.java [hunk 24]
```diff
@@ -31,17 +26,17 @@
...
     private static final   String NATIVE_TYPE_URL = "native";
 
-    protected Context             context;
-    protected Endpoint            endpoint;
-    protected Map<String, String> requestParameters = new HashMap<String, String>();;
-    protected Listener            listener;
+    protected Context  mContext;
+    protected Endpoint mEndpoint;
+    protected Map<String, String> requestParameters = new HashMap<String, String>();
+    protected Listener            mListener;
```
Comment (cerberillo) [inline library/src/main/java/net/pubnative/library/request/PubnativeRequest.java]: align parameters please

File: library/src/main/java/net/pubnative/library/request/PubnativeRequest.java [hunk 25]
```diff
@@ -31,17 +26,17 @@
...
     protected static final String BASE_URL        = "http://api.pubnative.net/api/partner/v2/promotions";
     private static final   String NATIVE_TYPE_URL = "native";
 
-    protected Context             context;
-    protected Endpoint            endpoint;
-    protected Map<String, String> requestParameters = new HashMap<String, String>();;
-    protected Listener            listener;
+    protected Context  mContext;
+    protected Endpoint mEndpoint;
+    protected Map<String, String> requestParameters = new HashMap<String, String>();
```
Comment (cerberillo) [inline library/src/main/java/net/pubnative/library/request/PubnativeRequest.java]: why `requestParameters` instead `mRequestParameters`?

File: library/src/test/java/net/pubnative/library/request/ListenerTest.java [hunk 26]
```diff
@@ -0,0 +1,86 @@
...
+    @Test
+    public void testWithNoListenerForSuccess() {
+
+        PubnativeRequest.Listener   listener        = spy(PubnativeRequest.Listener.class);
+        PubnativeRequest            request         = spy(new PubnativeRequest(this.applicationContext));
+
+        request.mListener = null;
+
+        request.invokeOnPubnativeRequestSuccess(mock(ArrayList.class));
+        verify(listener, times(0)).onPubnativeRequestSuccess(eq(request), any(List.class));
```
Comment (cerberillo) [inline library/src/test/java/net/pubnative/library/request/ListenerTest.java]: This test is useless, if you don't declare a listener you cannot check a listener xDD

For this type of thest the verify can be removed and the listener creation also, since we only want to ensure that `request.invokeOnPubnativeRequestSuccess(mock(ArrayList.class));` doesn't crash

Same for failure :)

**Comment by cerberillo** [comment]
Fixing this couple of errors, you can merge automatically once green light from CircleCI

**Comment by cerberillo** [comment]
Fix latest coments and merge :+1:

File: library/src/main/java/net/pubnative/library/network/PubnativeAPIRequest.java (hunk 21 above)
Comment (anshumansworld) [inline library/src/main/java/net/pubnative/library/network/PubnativeAPIRequest.java]: :+1:

File: library/src/main/java/net/pubnative/library/network/PubnativeAPIRequestTask.java (hunk 22 above)
Comment (anshumansworld) [inline library/src/main/java/net/pubnative/library/network/PubnativeAPIRequestTask.java]: :+1:

File: library/src/main/java/net/pubnative/library/request/PubnativeRequest.java (hunk 24 above)
Comment (anshumansworld) [inline library/src/main/java/net/pubnative/library/request/PubnativeRequest.java]: :+1:

File: library/src/test/java/net/pubnative/library/request/ListenerTest.java (hunk 26 above)
Comment (anshumansworld) [inline library/src/test/java/net/pubnative/library/request/ListenerTest.java]: :+1:

File: library/src/main/java/net/pubnative/library/request/PubnativeRequest.java (hunk 25 above)
Comment (anshumansworld) [inline library/src/main/java/net/pubnative/library/request/PubnativeRequest.java]: :+1:

File: library/src/main/java/net/pubnative/library/network/PubnativeAPIRequestTask.java (hunk 23 above)
Comment (anshumansworld) [inline library/src/main/java/net/pubnative/library/network/PubnativeAPIRequestTask.java]: :+1:
